ls1intum/Artemis
# Improve user service

This PR addresses a couple of issues with the user service, including a bug that is probably related to caching user entities.

### Checklist

### Motivation and Context
When users get instructor or tutor rights, they need to logout and login again, most probably because of caching issues.

### Description
I fixed some caching issues and refactored the code so that it is easier to understand and cleaner.
This also fixes some smaller issues when deleting users and adds a TODO to cleanup all user related 

### Steps for Testing
You need 2 users (A, B) and 1 course (C)
C: Course
A: Instructor in C
B: Normal user (who has not any instructor rights yet)

1. Log in to Artemis with both users in two different browsers
2. User A: Navigate to Course Administration and grant B instructor rights in C
3. User B: Refresh the page and check that you can see course administration now (without logout and login)
4. User A: Remove B from the instructors in C
5. User B: Refresh the page and check that you cannot see course administration now (without logout and login)

Please also test other related use cases:
* Course registration for students on TS1 and TS2 should work as before
* Login and logout should work as before
* Registering a new user account (TS2) should work as before
